### PR TITLE
Feature/Search and Replace

### PR DESCRIPTION
**Pull request by neCo2**
Adding a search and replace function. Currently accessible via a button below the editor preferences button.
The UI for search and replace uses a draggable component. The "active" surface for that is the darkened area with the title.
firefox_2024-06-07_21-15-25

Right now, it can do plaintext and regex replacements. I want to add template replacements too, but maybe I should wait for your chat completions PR for that one.

### CODE REVIEW

File: mikupad.html [hunk 1]
```diff
@@ -2014,6 +2070,161 @@
...
+			case 2:
+				templateReplace(search,replace,inputElement)
+				break;
+		}
+	}
+
+	function plaintextReplace(search,replace,elem) {
+		// console.log(elem.value)
+		// need to figure out a smart way to keep the cursor position
+		elem.value = elem.value.replaceAll(search,replace)
```
Comment (neCo2) [inline mikupad.html:2187]: And I could actually already use some help. In`SearchAndReplaceWidget` I tried to import promptArea, but I can't seem to use it in my functions unless I do weird stuff with useEffect a little further up.
Problem is that now, after replacing, the changes revert on predicting or reloading unless you do a manual edit first, and I don't have the slightest clue how to fix this.

Comment (lmg-anon) [reply mikupad.html]: I'm not sure why you can't import promptArea, but the issue with the changes reverting is simple to solve: You need to trigger the input event. You would have to either somehow call onInput like [this](https://github.com/lmg-anon/mikupad/blob/d5ba8962af85b618d4948ee42da7996eeec15d2c/mikupad.html#L3952), or trigger the input event like this:

```js
const event = new Event('input', { bubbles: true });
promptArea.current.dispatchEvent(event);
```"

Comment (neCo2) [reply mikupad.html]: Right, that works. Thanks!

**Comment by neCo2** [comment]
And that should be it except for the template substitution, which I'll add at a later point in time.
firefox_2024-06-08_11-23-47
Currently the replace button just replaces everything (unless you're using a non-global regex), so maybe a "replace only current match" button could be added at some point.
I'll also try to figure out how to add this to the undo stack, but feel free to do that yourself if you know a simple way.

**Comment by lmg-anon** [comment]
>I'll also try to figure out how to add this to the undo stack, but feel free to do that yourself if you know a simple way.

Theoretically, you shouldn't have to do anything, it will be handled as if the user replaced the text manually.

**Comment by neCo2** [comment]
Resolved the merge conflict and changed the svgs to components.
> Theoretically, you shouldn't have to do anything, it will be handled as if the user replaced the text manually.
Either I'm stupid, or it doesn't.

**Comment by lmg-anon** [comment]
>Either I'm stupid, or it doesn't.

Sorry, I mean after https://github.com/lmg-anon/mikupad/pull/66 is merged. The current undo stack only supports changes made at the end of the prompt.

**Comment by lmg-anon** [comment]
I got around to reviewing this PR, and the code looks good to me. However, I have some suggestions for the UI:

image

- Is the "Count" button really necessary? I think it would be fine to count as the user types.
- The "No matches found" warning also seems unnecessary; showing "0 Matches" would be enough.
- It's a bit odd to see "Find Next" but not "Find Last". I suggest not having the "Find Next" text.

Also, I see what you did with the placeholders, quite controversial lol. It's fine though, I won't pick on your creative choices.

**Comment by lmg-anon** [comment]
Okay, this looks good to merge, I think there are still a few things that could be improved, but I guess they could be done in other PRs as needed. Thank you for the contribution! :)

**Comment by neCo2** [comment]
@lmg-anon 
> I think there are still a few things that could be improved

Feel free to let me know what the issues are exactly.

**Comment by lmg-anon** [comment]
@neCo2 The only thing I would call an "issue" is how the widget resizes with the window size, becoming too big on an ultra-wide screen:
image

I guess this was my fault for suggesting `vw` for the widget size, probably `em` was the right call.

The other improvements I was considering include:
- Highlighting the matches like how other editors do.
- Showing the current match index/count.
- Opening the widget when you press CTRL+F or CTRL+H.
- Adding a "Replace Next" button.
- Adding a case sensitivity toggle.
- Adding an option to search/replace only within the selected text.

In other words, a very complete find/replace feature like in VSCode:
image
